### ISSUE TITLE
[Proposal] Route Registration Refactor

### ISSUE THREAD

**Issue by lindyhopchris** [cloudcreativity/laravel-json-api#53]
# Resource Route Changes

This change will modify route registration and remove the need for the `Request` class.

The new route registration will look like this:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'has-one' => ['author'],
        'has-many' => ['commments'],
    ]);
});
```

> Our intention is to remove the `Request` class to reduce the number of units that need to be generated per resource.
> We are moving its logic to middleware, and the properties on it that affect validation of query parameters are being moved to the `Validators` class.

## Controllers

The controller used for each resource will be the resource name plus `Controller`. For example, `posts` will be mapped to `PostsController`.

To override the controller, the following is used:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'controller' => 'SomeOtherController',
        'has-one' => ['author'],
        'has-many' => ['commments'],
    ]);
});
```

## Authorizers

The route register will automatically register the JSON API `Authorizer` to use for the resource type.
This will be based on the resource name and the `namespace` and `by-resource` settings in the `json-api` config file.

For example, with a namespace of `App\JsonApi` and `by-resource` as `true` the following class will be used as the authorizer if it exists:
`App\JsonApi\Posts\Authorizer`.

If `by-resource` is `false` then this class would be used:
`App\JsonApi\Authorizers\Posts`

To specify an authorizer that is different from this logic:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'authorizer' => \App\JsonApi\GenericAuthorizer::class,
        'has-one' => ['author'],
        'has-many' => ['commments'],
    ]);
});
```

## Validators

These will also be registered using the same logic as described for the authorizer.

For example, with a namespace of `App\JsonApi` and `by-resource` as `true` the following class will be used as the validators if it exists:
`App\JsonApi\Posts\Validators`.

If `by-resource` is `false` then this class would be used:
`App\JsonApi\Validators\Posts`

To specify validators that are different from this logic:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'validators' => \App\JsonApi\GenericValidators::class,
        'has-one' => ['author'],
        'has-many' => ['commments'],
    ]);
});
```

## Id Constraints

To add an id constraint for a particular resource:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'has-one' => ['author'],
        'has-many' => ['commments'],
        'id' => '[1-9]{1}[0-9]*'
    ]);
});
```

> To add an id constraint for all resources, a global pattern can already be registered for `resource_id`.

## Internal Implementation

This:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    JsonApi::resource('posts', [
        'has-one' => ['author'],
        'has-many' => ['commments'],
    ]);
});
```

Is equivalent to this:

```php
Route::group(['middleware' => 'json-api:v1'], function () {
    Route::group(['prefix' => '/posts', 'middleware' => [
        'json-api.authorize:\App\JsonApi\Posts\Authorizer',
        'json-api.validate:\Appi\JsonApi\Posts\Validators',
    ]], function () {
        Route::get('/', 'PostsController@index');
        Route::post('/', 'PostsController@create');
        Route::get('/{resource_id}', 'PostsController@read');
        Route::patch('/{resource_id}', 'PostsController@update');
        Route::delete('/{resource_id}', 'PostController@delete');

        // has-one and has-many:
        Route::get('/{resource_id}/{relationship_name}', 'PostsController@readRelatedResource')->where('relationship_name', 'author|comments');
        Route::get('/{resource_id}/relationships/{relationship_name}', 'PostsController@readRelationship')->where('relationship_name', 'author|comments');
        Route::patch('/{resource_id}/relationships/{relationship_name}', 'PostsController@replaceRelationship')->where('relationship_name', 'author|comments');
        // has-many only:
        Route::post('/{resource_id}/relationships/{relationship_name}', 'PostsController@addToRelationship')->where('relationship_name', 'comments');
        Route::delete('/{resource_id}/relationships/{relationship_name}', 'PostsController@removeFromRelationship')->where('relationship_name', 'comments');
    });
});
```

**The above also needs to push the `resource_type` param into the defaults for each route so that the request interpreter knows which resource the request relates to.**

> Note that the `json-api.authorizer` and `json-api.validators` middleware will only be added to the group if those classes exist.

**Comment by lindyhopchris** [comment]
ping @jstoone

**Comment by lindyhopchris** [comment]
See also #56 

The routing re-design needs to not register methods that aren't supported for the resource. E.g. if you cannot create or update a resource, the `POST`/`PATCH` routes should not be registered.

**Comment by lindyhopchris** [comment]
Merged into `develop` and will be in v0.8
